sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowFunctionFrame.scala

```diff
@@ -284,9 +284,7 @@ final class UnboundedWindowFunctionFrame(
 
   /** Write the frame columns for the current row to the given target row. */
   override def write(index: Int, current: InternalRow): Unit = {
-    // Unfortunately we cannot assume that evaluation is deterministic. So we need to re-evaluate
-    // for each row.
-    if (processor != null) {
+    if (processor != null && (index == 0)) {
```

Comment:
I can't remember why I added this. Are you sure that computing the result of an aggregate (on its buffer) is always deterministic? If so then I am ok with this change.

Reply:
so unbounded window frame means we only need to calculate result once and always return the result for each input row?

Then shall we call `processor.evaluate(target)` in `prepare` and do thing in `write`?

Reply:
That would also work :)

Reply:
> I can't remember why I added this. Are you sure that computing the result of an aggregate (on its buffer) is always deterministic? If so then I am ok with this change.

I've checked the subclass of `AggregateFunction`, and didn't found non-deterministic agg. 

Logically speaking, only non-ordered window spec is resolved to unbounded window frame (https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala#L2719), so if `processor.evaluate(target)` is various for each row in a same group, the result is always non-deterministic, right?